### PR TITLE
feat: set content length and other http headers so clients can show p…

### PR DESCRIPTION
**Pull request by tonsV2**
…rogress etc.

### CODE REVIEW

File: pkg/database/handler.go [hunk 1]
```diff
@@ -437,15 +437,20 @@ func (h Handler) Download(c *gin.Context) {
 		return
 	}
 
-	download, err := h.databaseService.Download(uint(id))
+	contentLength, reader, err := h.databaseService.Download(uint(id))
```
[the remark this answers is not in the document]
Comment (teleivo) [reply pkg/database/handler.go:440]: I would personally create higher level abstractions on top of lets the s3 client. I would push the responsibility of writing/copying down into it. So passing in the destination like you do to io.Copy.

Something like

`Download(id int, dst io.Writer) (int64, error)`

this also solves your triplet question 😋

File: pkg/database/handler.go [hunk 2]
```diff
@@ -437,15 +437,20 @@ func (h Handler) Download(c *gin.Context) {
...
 
 	_, file := path.Split(d.Url)
 	c.Header("Content-Disposition", "attachment; filename="+file)
-	_, err = io.Copy(c.Writer, download)
+	c.Header("Content-Description", "File Transfer")
+	c.Header("Content-Transfer-Encoding", "binary")
+	c.Header("Content-Type", "application/octet-stream")
+	c.Header("Content-Length", fmt.Sprintf("%d", contentLength))
+
+	_, err = io.Copy(c.Writer, reader)
```
[the remark this answers is not in the document]
Comment (teleivo) [reply pkg/database/handler.go:453]: Does gin close the response body? Just want to make sure 😄

File: pkg/database/service.go [hunk 3]
```diff
@@ -105,19 +105,19 @@ func (s service) Upload(d *model.Database, group *models.Group, file io.Reader,
...
+func (s service) Download(id uint) (int64, io.ReadCloser, error) {
 	d, err := s.repository.FindById(id)
 	if err != nil {
-		return nil, err
+		return 0, nil, err
 	}
 
 	if d.Url == "" {
-		return nil, apperror.NewBadRequest(fmt.Sprintf("database with %d doesn't reference any url", id))
+		return 0, nil, apperror.NewBadRequest(fmt.Sprintf("database with %d doesn't reference any url", id))
```
Comment (teleivo) [inline pkg/database/service.go:115]: The package says apperror but the method sounds like an HTTP 400 status. I would keep HTTP (bad request) out of app errors and services, return an error here and let the handler decide if thats a 4xx or else. In case we want to use the services for something else than HTTP.

**Review by teleivo — COMMENTED** [review]
Made some suggestions 😄

File: pkg/database/handler.go [hunk 4]
```diff
@@ -443,20 +442,15 @@ func (h Handler) Download(c *gin.Context) {
...
 	_, file := path.Split(d.Url)
-	c.Header("Content-Disposition", "attachment; filename="+file)
-	c.Header("Content-Description", "File Transfer")
-	c.Header("Content-Transfer-Encoding", "binary")
-	c.Header("Content-Type", "application/octet-stream")
-	c.Header("Content-Length", fmt.Sprintf("%d", contentLength))
-
-	_, err = io.Copy(c.Writer, reader)
+	cb := func(contentLength int64) {
+		c.Header("Content-Disposition", "attachment; filename="+file)
```
Comment (teleivo) [inline pkg/database/handler.go:447]: the headers that are not related to content length could move outside of the callback. this way you could also inline the callback into the function call to Download using an anonymous function (spares us the pain of having to name 😋 ).

**Review by teleivo — APPROVED** [review]
nice 🎉 some minor ideas. 😋

File: pkg/database/handler.go [hunk 5]
```diff
@@ -443,20 +442,15 @@ func (h Handler) Download(c *gin.Context) {
...
+		c.Header("Content-Description", "File Transfer")
+		c.Header("Content-Transfer-Encoding", "binary")
+		c.Header("Content-Type", "application/octet-stream")
+		c.Header("Content-Length", fmt.Sprintf("%d", contentLength))
```
[the remark this answers is not in the document]
Comment (tonsV2) [reply pkg/database/handler.go]: I've updated

File: pkg/database/handler.go [hunk 6]
```diff
@@ -443,20 +442,15 @@ func (h Handler) Download(c *gin.Context) {
...
+	}
+	err = h.databaseService.Download(uint(id), c.Writer, cb)
```
[the remark this answers is not in the document]
Comment (tonsV2) [reply pkg/database/handler.go]: I planned on doing a project wide search for strconv.Atoi and replacing with strconv.ParseUint. If that's what you're referring to?
